File: Coin_Change/coin_change.cpp
#include <iostream>
#include <vector>
using namespace std;

int num_ways( const vector <int>& coins, const int& num_cents )
{
    vector <int> num_ways( num_cents + 1, 0 );
    num_ways[0] = 1;

    for( int i = 0; i < coins.size(); i++ )
        for( int j = 0; j + coins[i] <= num_cents; j++ )
            num_ways[j + coins[i]] += num_ways[j];

    return num_ways[num_cents];
}

int main()
{
    int num_coins, num_cents;
    cin >> num_coins >> num_cents;

    vector <int> coins( num_coins );
    for( int i = 0; i < num_coins; i++ )
        cin >> coins[i];

    cout << "There are " << num_ways( coins, num_cents ) << " ways to make change for " << num_cents << " cents." << endl;

    return 0;
}

File: Ad_Hoc/Coin_Change/coin_change.java
import java.util.Scanner;

class coin_change
{
    public static int get_total_num_coins( int[] coin_values, final int num_cents )
    {
        int[] num_ways_make_change = new int[num_cents + 1];
        num_ways_make_change[0] = 1;

        for( int i = 0; i < coin_values.length; i++ )
            for( int j = 0; j + coin_values[i] <= num_cents; j++ )
                num_ways_make_change[j + coin_values[i]] += num_ways_make_change[j];

        return num_ways_make_change[num_cents];
    }

    public static void main( String[] args )
    {
        Scanner in = new Scanner( System.in );
        int num_coins = in.nextInt();
        int num_cents = in.nextInt();
        int[] coin_values = new int[num_coins];

        int tmp;
        for( int i = 0; i < num_coins; i++ )
        {
            tmp = in.nextInt();
            coin_values[i] = tmp;
        }

        System.out.println( "Change for " + num_cents + " cents can be made in " + get_total_num_coins( coin_values, num_cents ) + " ways." );
    }
};
